UIPS/UIP-0122.md

```diff
@@ -0,0 +1,156 @@
+
+A separate rune (`~.` is proposed) can extract core labels from the type of an axis of the subject and only needs to be wrapped around *outer entry points* (e.g. the outer arms of Arvo.) This composes the `%wild` hint from the Hoon type of the given axis of the subject.
+
+Thus, no relabeling of cores is necessary, the only Hoon level change is the addition of a simple rune wrapping a fixed number of entry points to Arvo. Hoon can compose the `%wild` hint's clue entirely from information tracked in the subject type.
+
+## Example
+
+Consider the following jetted standard library (`vial.hoon`)
+
+```hoon=
+=>  ~%  %vial.100  %vial
+~%  %one  +3..  ~
+|%
+~/  %dec
+++  dec
+  |=  a=@
+  ?<  =(a 0)
+  =|  b=@
+  |-  ^-  @
+  ?:  =(a +(b))  b
+  $(b +(b))
+--
+```
+
+The `%fast` hints on lines 1, 2, and 4 are compiled into Nock, but also result in labels in the Hoon type. For the `+dec` gate, the core is labeled `/dec/one/vial.100`.
+
+Now consider the following trivial lifecycle door (`life.hoon`):
+
+
+```hoon=
+/+  vial
+=>  vial
+|_  many=@
+++  this  .
+++  peek
+  |=  pax=*
+  ?>  =(pax /how/<many>)
+  many
+++  poke
+  |=  up=?
+  ~.  +7..
+  ^-  _this
+  ?:  up  this(many +(many))
+  this(many (dec many))
+--
+```
+  
+In this example, the `~.` rune would compile to a `%wild` hint with three elements:
+
+- `[l=/vial.100 s=[& %vial]]`
+- `[l=/one/vial.100 s=[[& &] [<one battery> %vial]]`
+- `[l=/dec/one/vial.100 s=[[& | & &] <dec battery> ~ <one battery> %vial]`
+
+(In practice the `cape`s would be normalized by rewriting `[& &]` to `&`, however here we leave them denormalized for ease of following the example.)
+
+For the entire scope of the `~.` rune, a conforming interpreter would use this as the set of registered cores in order to match jets or produce traces. The list is extracted labels from the *type* of the first subhoon: `+7..` and (recursively) the result types of any arms of any cores in the type.
+
+## Bonus: fix `+mink` jet mismatch
+
+A further advantage, pointed out by ~master-morzod, is that type-level availability of core labels could be used to fix the jet mismatch of `+mink` in an ergonomic way. If the sample of `+mink` is extended to contain jet registrations (in the same `$wilt` mold), its semantics can be fixed to never produce a stack trace underneath a labeled core. The jet for `+mink` can load the supplied `$wilt` as its local jet registrations, and the interpreter can likewise refuse to produce stack traces under labeled cores, even if they are in fact unjetted. Of course `+mink` should also explicitly handle the `%wild` hint.
+
+A vase-mode wrapper around `+mink` can then extract the necessary registrations from the type carried by the vase, using the same algorithm as for compiling `~.`, but at runtime. The extraction would almost certainly need to be persistently memoized for reasonable performance.
```

Comment:
> extract registrations from the type carried by the vase

Which vase is this referring to?  The arguments to `+mink` don't have any varying types: there's the untyped subject and formula to run, and then the added `$wilt` argument, which has a static type.  So what type are the registrations pulled from?

I assume the compiler will want to pull them from some subject type somewhere, but by the time we arrive at a call site for `+mink`, the subject is the `+mink` gate itself... so I think we would need to wrap a containing expression.

Or is the idea that you pass a subject vase into `+mink`, instead of an untyped subject noun?  I like the idea of statically generating a type and code-gen'ing a `$wilt` data structure to be passed to `+mink`, but I'm struggling to think of what mechanism you have in mind for that.

Reply:
Without new syntactic sugar, I would imagine doing something like this (where `+extract-registrations` would be in hoon.hoon):
```hoon
=|  state=@
|%
++  this  .
++  run
  ^-  *
  =/  subject  dec
  =/  formula  =>(dec !=((dec 33)))
  =/  =wilt    ^~((extract-registrations -:!>(subject)))
  (mink [subject formula] wilt)
--
```

As long as `subject` can be constant-folded at compile time, the `^~` should be able to statically compute the registrations based on the type.  I don't know what kind of sugar we might want for this.